torch/csrc/distributed/c10d/init.cpp

```diff
@@ -2404,6 +2404,26 @@ options :class:`~torch.distributed.ProcessGroupNCCL.Options`).
       "_get_intra_node_comm_usage_counter",
       &::c10d::intra_node_comm::getIntraNodeCommUsageCounter);
 
+  using IntraNodeComm = ::c10d::intra_node_comm::IntraNodeComm;
+  py::class_<IntraNodeComm, c10::intrusive_ptr<IntraNodeComm>>(
+      module, "_IntraNodeComm")
+      .def_static(
+          "rendezvous",
+          &IntraNodeComm::rendezvous,
+          py::arg("store"),
+          py::arg("prefix"),
+          py::arg("rank"),
+          py::arg("world_size"),
+          py::arg("buffer_size") = 0)
+      .def("barrier", &IntraNodeComm::barrier, py::arg("ranks") = py::none())
```

Comment:
it seems that the `ranks` are not used anywhere? do you for see that there would be cases that we only `barrier` in certain set of ranks? iiuc maybe we should always barrier on all ranks for this intra node communicator to align with ProcessGroup::barrier?

Reply:
> it seems that the ranks are not used anywhere?

This is used. The barrier in this PR supports synchronization among a subset of members.

> do you for see that there would be cases that we only barrier in certain set of ranks?

For example, to achieve efficient all-gather matmul on a host without NVSwitch but with a ring topology, a user can leverage finer-grained synchronization to create their own protocol.

> iiuc maybe we should always barrier on all ranks for this intra node communicator to align with ProcessGroup::barrier?

This will be the behavior when we expose this as `ProcessGroup::broad`. The finer-grained barrier will be exposed via different interfaces.